src/participant.rs

```diff
@@ -0,0 +1,218 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use ed25519_dalek::Signature;
+use ed25519_dalek::SignatureError;
+use ed25519_dalek::Signer;
+use ed25519_dalek::SigningKey;
+use ed25519_dalek::Verifier;
+use ed25519_dalek::VerifyingKey;
+use rand_core::CryptoRng;
+use rand_core::RngCore;
+use std::cell::OnceCell;
+use std::io;
+use x25519_dalek::PublicKey;
+use x25519_dalek::StaticSecret;
+
+const VERIFICATION_KEY_LEN: usize = ed25519_dalek::PUBLIC_KEY_LENGTH;
+const ENCRYPTION_KEY_LEN: usize = 32;
+const SIGNATURE_LEN: usize = Signature::BYTE_SIZE;
+const VERSION_LEN: usize = 1;
+const VERSION: [u8; VERSION_LEN] = [0x72];
+
+pub const IDENTITY_LEN: usize =
+    VERSION_LEN + VERIFICATION_KEY_LEN + ENCRYPTION_KEY_LEN + SIGNATURE_LEN;
+
+pub type IdentitySerialization = [u8; IDENTITY_LEN];
+
+/// Returns the portion of identifier data that is signed by [`Secret::signing_key`]
+fn authenticated_data(
+    verification_key: &VerifyingKey,
+    encryption_key: &PublicKey,
+) -> [u8; VERSION_LEN + VERIFICATION_KEY_LEN + ENCRYPTION_KEY_LEN] {
+    let mut data = [0u8; VERSION_LEN + VERIFICATION_KEY_LEN + ENCRYPTION_KEY_LEN];
+    let parts = [
+        &VERSION[..],
+        verification_key.as_bytes(),
+        encryption_key.as_bytes(),
+    ];
+    let mut slice = &mut data[..];
+    for part in parts {
+        slice[..part.len()].copy_from_slice(part);
+        slice = &mut slice[part.len()..];
+    }
+    assert_eq!(slice.len(), 0);
+    data
+}
+
+/// Secret keys of a participant.
+#[derive(Clone)]
+pub struct Secret {
+    signing_key: SigningKey,
+    decryption_key: StaticSecret,
+    // Use a cell to lazily compute (and cache) the `Identity` corresponding to this `Secret`, to
+    // avoid computing public keys and signatures multiple times. This makes `Secret` not `Sync`,
+    // but this can be revisited if it turns out to be a problem.
+    identity: OnceCell<Identity>,
+}
+
+impl Secret {
+    #[must_use]
+    pub fn random<T: RngCore + CryptoRng>(mut csprng: T) -> Self {
+        Self {
+            signing_key: SigningKey::generate(&mut csprng),
+            decryption_key: StaticSecret::random_from_rng(&mut csprng),
+            identity: OnceCell::new(),
+        }
+    }
+
+    #[must_use]
+    pub fn to_identity(&self) -> Identity {
```

Comment:
Do you want to include an implicit `into` method here that uses this?

Reply:
Good point; will add all the relevant `From` impls